pkg/store/bucket.go

```diff
@@ -1188,89 +1150,200 @@ func newBucketIndexReader(ctx context.Context, logger log.Logger, block *bucketB
+	}
 
-		msg := fmt.Sprintf("found parts: %v bases on:", parts)
-		for _, p := range ps {
-			msg += fmt.Sprintf(" [start: %v; end: %v]", p.ptr.Start, p.ptr.End)
+	// As of version two all series entries are 16 byte padded. All references
+	// we get have to account for that to get the correct offset.
+	// We do it right at the beginning as it's easier than doing it more fine-grained
+	// at the loading level.
+	if r.block.indexVersion >= 2 {
+		for i, id := range ps {
+			ps[i] = id * 16
 		}
-
-		return errors.Errorf("expected all postings to be loaded but spotted malformed one with key: %v; start: %v; end: %v. Additional info: %s",
-			postings.key, postings.ptr.Start, postings.ptr.End, msg)
 	}
 
-	return nil
+	return ps, nil
 }
 
-// loadPostings loads given postings using given start + length. It is expected to have given postings data within given range.
-func (r *bucketIndexReader) loadPostings(ctx context.Context, postings []*lazyPostings, start, end int64) error {
-	begin := time.Now()
+// NOTE: Derived from tsdb.postingsForMatcher. index.Merge is equivalent to map duplication.
+func matchingLabels(lvalsFn func(name string) []string, m labels.Matcher) (labels.Labels, error) {
+	// If the matcher selects an empty value, it selects all the series which don't
+	// have the label name set too. See: https://github.com/prometheus/prometheus/issues/3575
+	// and https://github.com/prometheus/prometheus/pull/3578#issuecomment-351653555
+	if m.Matches("") {
+		// We don't support tsdb.postingsForUnsetLabelMatcher.
+		// This is because it requires fetching all postings for index.
+		// This requires additional logic to avoid fetching big bytes range (todo: how big?). See https://github.com/prometheus/prometheus/pull/3578#issuecomment-351653555
+		// to what it blocks.
+		return nil, errors.Errorf("support for <> != <val> matcher is not implemented; empty matcher for label name %s", m.Name())
+	}
 
-	b, err := r.block.readIndexRange(ctx, int64(start), int64(end-start))
-	if err != nil {
-		return errors.Wrap(err, "read postings range")
+	// Fast-path for equal matching.
+	if em, ok := m.(*labels.EqualMatcher); ok {
+		return labels.Labels{{Name: em.Name(), Value: em.Value()}}, nil
 	}
 
-	r.mtx.Lock()
-	defer r.mtx.Unlock()
+	var matchingLabels labels.Labels
+	for _, val := range lvalsFn(m.Name()) {
+		if m.Matches(val) {
+			matchingLabels = append(matchingLabels, labels.Label{Name: m.Name(), Value: val})
+		}
+	}
 
-	r.stats.postingsFetchCount++
-	r.stats.postingsFetched += len(postings)
-	r.stats.postingsFetchDurationSum += time.Since(begin)
-	r.stats.postingsFetchedSizeSum += int(end - start)
+	return matchingLabels, nil
+}
 
-	for _, p := range postings {
-		c := b[p.ptr.Start-start : p.ptr.End-start]
+type postingPtr struct {
+	key labels.Label
+	ptr index.Range
+}
 
-		_, l, err := r.dec.Postings(c)
-		if err != nil {
-			return errors.Wrap(err, "read postings list")
+// fetchPostings returns sorted slice of postings.
```

Comment:
`fetchPostings returns sorted slice of postings that match the selected labels.`